aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -0,0 +1,399 @@
+  int64_t stride = 1;
+  for (int64_t d = dim; d < tensor.dim(); ++d) {
+    stride *= tensor.sizes()[d];
+  }
+  return stride * tensor.element_size();
+}
+
+// Calculate the number of chunks. This is the same for all splits.
+static inline int64_t get_num_chunks(const at::Tensor& tensor, int64_t dim) {
+  int64_t num_chunks = tensor.numel();
+  for (int64_t d = dim; d < tensor.dim(); ++d) {
+    num_chunks /= tensor.sizes()[d];
+  }
+  return num_chunks;
+}
+
+// Pack multiple std::vector<int64_t> into a single cuda tensor.
+std::pair<at::Tensor, std::vector<int64_t*>> pack_vecs(
+    std::vector<const std::vector<int64_t>*> vecs,
+    const at::Device& device) {
+  int64_t numel = 0;
+  for (const auto* vec : vecs) {
+    numel += vec->size();
+  }
+
+  auto packed = at::empty(
+      {numel}, at::TensorOptions().dtype(at::kLong).pinned_memory(true));
+  size_t offset = 0;
+  for (const auto* vec : vecs) {
+    memcpy(
+        packed.data_ptr<int64_t>() + offset,
+        vec->data(),
+        sizeof(int64_t) * vec->size());
+    offset += vec->size();
+  }
+  packed = packed.to(device, /*non_blocking=*/true);
+
+  std::vector<int64_t*> ptrs;
+  ptrs.reserve(vecs.size());
+  offset = 0;
+  for (const auto* vec : vecs) {
+    ptrs.push_back(packed.data_ptr<int64_t>() + offset);
+    offset += vec->size();
+  }
+  return std::make_pair(std::move(packed), std::move(ptrs));
+}
+
+} // namespace detail
+
+// See [CUDA fast path for split_with_sizes_copy.out]
+void split_with_sizes_copy_out_cuda_contiguous_no_cast(
+    const at::Tensor& self,
+    at::IntArrayRef split_sizes,
+    int64_t dim,
+    at::TensorList out) {
+  const auto device = self.device();
+  const auto src_base_addrs =
+      detail::get_split_base_addrs(self, split_sizes, dim);
+  const auto dst_base_addrs = detail::get_dst_addrs(out);
+  const auto src_stride = detail::get_chunk_stride(self, dim);
+  const auto split_chunk_sizes =
+      detail::get_split_chunk_sizes(self, split_sizes, dim);
+  const auto num_chunks = detail::get_num_chunks(self, dim);
+
+  // Calculate the number of blocks required for the first chunk across all
+  // splits, assuming each thread only processes BYTES_PER_THREAD bytes.
+  int64_t num_blocks = 0;
+  for (const auto& split_chunk_size : split_chunk_sizes) {
+    num_blocks += detail::div_up(
+        split_chunk_size, detail::BLOCK_SIZE * detail::BYTES_PER_THREAD);
+  }
+
+  // Calculate the maximum number of blocks to launch. Only consider
+  // maxThreadsPerMultiProcessor as a limiting factor as the kernel uses no
+  // shared memory and little registers. Over-subscribe the SMs to hide I/O
+  // latency.
+  const auto num_sms =
+      at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+  const auto max_threads_per_sm =
+      at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor;
+  const int64_t max_blocks =
+      num_sms * max_threads_per_sm / detail::BLOCK_SIZE * 2.0;
```

Comment:
Does the oversubscription here come from the `* 2.0`? I.e., we launch a grid with 2x the max number of blocks compared to what can run simultaneously on the GPU?

Reply:
Exactly. The purpose is to achieve high sustained occupancy. If we launch just enough blocks that can be executed simultaneously, if some blocks finish early, the resource previously occupied by them will remain unoccupied until the kernel finishes, causing a drop in occupancy.

The source of workload imbalance comes from the small shards/chunks and the tail of large large shards/chunks.